k4MarlinWrapper/src/components/StoreUtils.cpp

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2019-2024 Key4hep-Project.
+ *
+ * This file is part of Key4hep.
+ * See https://key4hep.github.io/key4hep-doc/ for further info.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "GaudiKernel/AlgTool.h"
+#include "GaudiKernel/AnyDataWrapper.h"
+#include "GaudiKernel/IDataManagerSvc.h"
+#include "GaudiKernel/IDataProviderSvc.h"
+#include "GaudiKernel/SmartDataPtr.h"
+
+#include "StoreUtils.h"
+#include "podio/Frame.h"
+
+#include "k4FWCore/FunctionalUtils.h"
+
+#include <string>
+#include <vector>
+
+std::vector<std::string> getAvailableCollectionsFromStore(const AlgTool*                                  thisClass,
+                                                          std::optional<std::map<uint32_t, std::string>>& idToName,
+                                                          bool returnFrameCollections) {
+  std::vector<std::string> collectionNames;
+
+  SmartIF<IDataManagerSvc> mgr;
+  mgr = thisClass->evtSvc();
+
+  SmartDataPtr<DataObject> root(thisClass->evtSvc(), "/Event");
+  if (!root) {
+    thisClass->error() << "Failed to retrieve root object /Event" << endmsg;
+  }
+
+  auto pObj = root->registry();
+  if (!pObj) {
+    thisClass->error() << "Failed to retrieve the root registry object" << endmsg;
+  }
```

Comment:
Should we exit early here, or throw an exception to avoid running into a seg-fault? Or does `error()` already do that in this case?

Reply:
No, `error()` doesn't, however I don't know if there are any circumstances under which `error()` will run. The same code can be found in the Writer from k4FWCore and I have never seen it happening. I can add one `throw` if you really want, but then all the checks in that function should also have a `throw`.

Reply:
Does a `fatal` throw or abort execution? The main potential issue I see here is that if these checks fail we will have a `nullptr` access and a segfault, and depending on if / when stdout / stderr has last been flushed we may or may not see the error message. If there are no visible error messages in case things fail, these checks are almost useless, I think. Obviously, if the messages are guaranteed to be there even in case of a seg fault, then there is at least some information where things went wrong.

Reply:
`fatal` doesn't abort. `endmsg` writes the message so there should always be some information and it would happen right before the crash. As I said, I've never seen this fail because I think if it's in a state where that would fail something else will fail much earlier, like not being able to find `/Event` in the store. I'll just add the 3 or 4 throws since it's easier than continuing this conversation.